econml/dml.py

```diff
@@ -299,9 +349,11 @@ def coef_(self):
         return self._model_final.coef_
 
 
-class DMLCateEstimator(_DMLCateEstimatorBase):
+class LinearDMLCateEstimator(DMLCateEstimator):
```

Comment:
The following piece of code, returns a warning that I think we should resolve and also the performance of the intervals is terrible (even though 'statsmodels' inference is good). The warning that is returned during the effeEven more weirdly
```python
def exp_te(x):
    return np.exp(2*x[0])
np.random.seed(1234)
n=1000
n_w=30
support_size=5
n_x=1
support_Y=np.random.choice(range(n_w),size=support_size,replace=False)
coefs_Y=np.random.uniform(0,1,size=support_size)
epsilon_sample=lambda n:np.random.uniform(-1,1,size=n)
support_T=support_Y
coefs_T = np.random.uniform(0, 1, size=support_size)
eta_sample = lambda n: np.random.uniform(-1, 1, size=n) 
W = np.random.normal(0, 1, size=(n, n_w))
X = np.random.uniform(0, 1, size=(n, n_x))
TE = np.array([exp_te(x_i) for x_i in X])
log_odds = np.dot(W[:, support_T], coefs_T) + eta_sample(n)
T_sigmoid = 1/(1 + np.exp(-log_odds))
T = np.array([np.random.binomial(1, p) for p in T_sigmoid])
Y = TE * T + np.dot(W[:, support_Y], coefs_Y) + epsilon_sample(n)
X_test = np.array(list(product(np.arange(0, 1, 0.01), repeat=n_x)))
est = LinearDMLCateEstimator(model_y=Lasso(),
                             model_t=LogisticRegression(),
                             featurizer=PolynomialFeatures(degree=2, include_bias=True),
                             n_splits=2,
                             discrete_treatment=True)
from econml.inference import StatsModelsInference, BootstrapInference
est.fit(Y, T, X, W, inference=BootstrapInference(n_bootstrap_samples=100))
te_pred = est.effect(X_test)
te_pred_int = est.effect_interval(X_test, alpha=.01)
plt.figure(figsize=(10,6))
plt.plot(X_test,te_pred,label='DML estimate')
expected_te=np.array([exp_te(x_i) for x_i in X_test])
plt.plot(X_test,expected_te,'b--',label='True effect')
plt.fill_between(X_test[:, 0], te_pred_int[0], te_pred_int[1], alpha=.5)
plt.ylabel('Treatment Effect')
plt.xlabel('x')
plt.legend()
plt.show()
```

The warning is produced by the line:
`te_pred_int = est.effect_interval(X_test, alpha=.01)`
and it is:
`C:\ProgramData\Anaconda3\lib\site-packages\sklearn\preprocessing\label.py:252: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel(). y = column_or_1d(y, warn=True)`

Even more weirdly this warning is not produced when statsmodels inference is used. So it is something created by the bootstrap logic.

Reply:
@kbattocchi performance is now better, but the warning is still not resolved.

Reply:
actually one correction to my comment above: this warning is shown even when statsmodels inference is used.

Reply:
Here is were this is stemming from: at effect time, we create a vector T0 and T1 that is passed through the transformer. However, we create a vector of dimension (n_samples, d_t). But d_t=1 in the case of discrete treatment. However, the original shape of T was None, since it was a column vector. Then when we pass it through the transformer, the transformer contains a label encoder, which expects a vector and not a matrix. This is the exact code that throws the error:
```python
self.transformer = FunctionTransformer(
                func=(lambda T:
                      self._one_hot_encoder.transform(
                          reshape(self._label_encoder.transform(T), (-1, 1)))[:, 1:]),
                validate=False)
```

If for instance, here we replace it with 
```python
self.transformer = FunctionTransformer(
                func=(lambda T:
                      self._one_hot_encoder.transform(
                          reshape(self._label_encoder.transform(T.ravel()), (-1, 1)))[:, 1:]),
                validate=False)
```
the warning is fixed.

Reply:
@vasilismsr Actually it's worse than that - if there are more than 2 discrete treatments it really breaks, because d_t is overwritten to (n-1) so we expand the (unencoded) treatments to (n-1) columns in `_expand_treatments`, then try to encode after that.  Basically, for treatment expansion (used by effect, marginal_effect, and _interval versions in cate_estimator.py) we need to know the original dimensions of T in case we're given a scalar, but for generating unit treatments (used by const_marginal_effect and _interval version in dml.py and inference.py) we need to know the expanded dimensions.  We can't use self._d_t for both...

Reply:
For now I'll fix this in the next push by storing the original input inside the treatment expansion mixin (which is the component that needs the original input).  But I'd like to move away from using a mixin at all in the near future, since as you pointed out the mixin wraps its superclass's effect but can't wrap the effect method of the class that it's mixed into if it's also overridden there.  This can definitely be addressed by using a class decorator instead of a mixin, but I'd like to see if anyone's got any brighter ideas for reorganizing this logic.